### PR TITLE
add DiffInt

### PR DESCRIPTION
**Pull request by riaqn**
Some warning (constraint unsolved) remaining and I can't fix.

### CODE REVIEW

**Comment by mortberg** [comment]
I made a PR to your local branch fixing the yellow. Merge it in and then this should be ready

**Comment by riaqn** [comment]
I merged it.

**Comment by mortberg** [comment]
Can you please also add DiffInt to the Everything file in Data?

**Comment by riaqn** [comment]
There seems to be some glitch in my code - when I try to normalize `discreteℤ [(2 , 4)] [(3 , 5)]`, it gives me back `yes` of something, which is expected.

But when I normalize `discreteℤ [(2 , 4)] [(3 , 4)]`, agda just freezes and takes 100% CPU. It should give me `no` of something.

**Comment by 3abc** [comment]
> There seems to be some glitch in my code (or Agda) - when I try to normalize `discreteℤ [(2 , 4)] [(3 , 5)]`, it gives me back `yes` of something, which is expected.
> 
> But when I normalize `discreteℤ [(2 , 4)] [(3 , 4)]`, agda just freezes and takes 100% CPU. It should give me `no` of something.

Likely just an Agda's performance problem... Maybe `discreteSetQuotients` is too complicated for Agda to normalize.

**Comment by 3abc** [comment]
> Likely just an Agda's performance problem...

My suspicion is, even if you make some relatively simple equivalence relation on `Nat` like `m ~ n` iff `m - n` is even. It's going to hang if you prove `discrete Nat/~` and evaluate it on `[0]` `[1]`

**Comment by riaqn** [comment]
@3abc you are right. I tried an even more trivial one (that 0-0 = 0-1) and it still hangs.

However, I'm not convinced it's a performance problem than it's a BUG. I don't think the normalization involved is that complex.

**Comment by mortberg** [comment]
> @3abc you are right. I tried an even more trivial one (that 0-0 = 0-1) and it still hangs.
> 
> However, I'm not convinced it's a performance problem than it's a BUG. I don't think the normalization involved is that complex.

Are you tests the same as:

https://github.com/mortberg/cubicaltt/blob/master/examples/hz.ctt#L63

?

Computing `0=1` and `1=1` is instant:
```
> test01
EVAL: false
Time: 0m0.089s
> test11
EVAL: true
Time: 0m0.093s
```
So this sounds like either a bug or that your proofs are too complicated.

File: Cubical/Relation/Binary/Base.agda [hunk 1]
```diff
@@ -9,13 +9,13 @@ open import Cubical.HITs.SetQuotients.Base
 
 module BinaryRelation {ℓ ℓ' : Level} {A : Set ℓ} (R : A → A → Set ℓ') where
   isRefl : Set (ℓ-max ℓ ℓ')
-  isRefl = {a : A} → R a a
```
Comment (Alizter) [inline Cubical/Relation/Binary/Base.agda]: What is the reason for changing these?

Comment (mortberg) [reply Cubical/Relation/Binary/Base.agda]: Agda couldn't infer the argument in DiffInt.Properties so things got yellow there. I believe that we often want these arguments explicitly when doing these proofs, so I'd rather have them explicit and insert underscores sometimes.

**Comment by 3abc** [comment]
> However, I'm not convinced it's a performance problem than it's a BUG. I don't think the normalization involved is that complex.

Could be a bug too. But given that used `J` twice and `relIsTrans` contains too much `hcomp` the normal form will just explode for Agda to normalize...

Here is an example, I defined a really simple relation `rel2` below and asked Agda to normalize `discreteN/2 [ 0 ] [ 1 ]` it took 35.8 seconds of 100% CPU to produce a normal form of `42148` lines. In your case the normal form could be easily more than 4 million lines. (Now you might have some idea why in the other thread we want things to be definitionally equal badly, and try to avoid using `J`.)

```Agda
open import Cubical.Data.Empty public

data ⊤ : Set where
  t : ⊤

rel2 : ℕ → ℕ → Set
rel2 0 0 = ⊤
rel2 0 (suc n) = ⊥
rel2 (suc n) 0 = ⊥
rel2 (suc n) (suc m) = ⊤

N/2 = ℕ / rel2

rel2IsProp : BinaryRelation.isPropValued rel2
rel2IsProp zero zero t t = refl
rel2IsProp zero (suc b) x y = ⊥-elim x
rel2IsProp (suc a) zero x y = ⊥-elim x
rel2IsProp (suc a) (suc b) t t  = refl

rel2IsEquiv : isEquivRel rel2
rel2IsEquiv = EquivRel {A = ℕ} rel2IsRefl rel2IsSym rel2IsTrans
  where
    rel2IsRefl : isRefl rel2
    rel2IsRefl zero = t
    rel2IsRefl (suc a) = t

    rel2IsSym : isSym rel2
    rel2IsSym zero zero t = t
    rel2IsSym zero (suc b) x = ⊥-elim x
    rel2IsSym (suc a) zero x = ⊥-elim x
    rel2IsSym (suc a) (suc b) t = t
    
    rel2IsTrans : isTrans rel2
    rel2IsTrans zero zero c x y = y
    rel2IsTrans zero (suc b) c x _ = ⊥-elim x
    rel2IsTrans (suc a) zero c x _ = ⊥-elim x
    rel2IsTrans (suc a) (suc b) 0 x y = ⊥-elim y
    rel2IsTrans (suc a) (suc b) (suc c) t t = t

N/2IsDiscrete : (a b : ℕ) → Dec (rel2 a b)
N/2IsDiscrete zero zero = yes t
N/2IsDiscrete zero (suc b₁) = no λ x → x
N/2IsDiscrete (suc a₁) zero = no λ x → x
N/2IsDiscrete (suc a₁) (suc b₁) = yes t

discreteN/2 : Discrete N/2
discreteN/2 = discreteSetQuotients discreteℕ rel2IsProp rel2IsEquiv N/2IsDiscrete
```

**Comment by riaqn** [comment]
@3abc thanks for the example. That's indeed suprising to me; I didn't expect such a simple proposition would need such a big proof.

Could you post the normalized form somewhere so I can look at it?

**Comment by 3abc** [comment]
> Could you post the normalized form somewhere so I can look at it?

Won't be too hard for you to just normalize it LoL (40 seconds!). It's so long that when I use `M-e` to jump to the end emacs hangs for a few seconds... And I've no idea how to export.

**Comment by riaqn** [comment]
ahh, ok I got it.

yeah, this looks really bad. But I assume in general, if the type of the term is some inductive type, then by canonicity, the normalized form should be very short? Also I guess the process shouldn't take long because the term would be kept short?

**Comment by riaqn** [comment]
> > @3abc you are right. I tried an even more trivial one (that 0-0 = 0-1) and it still hangs.
> > However, I'm not convinced it's a performance problem than it's a BUG. I don't think the normalization involved is that complex.
> 
> Are you tests the same as:
> 
> https://github.com/mortberg/cubicaltt/blob/master/examples/hz.ctt#L63
> 
> ?
> 
> Computing `0=1` and `1=1` is instant:
> 
> ```
> > test01
> EVAL: false
> Time: 0m0.089s
> > test11
> EVAL: true
> Time: 0m0.093s
> ```
> So this sounds like either a bug or that your proofs are too complicated.

I believe there is difference between just saying "yes/no" and saying "yes/no" with a reason. The reason is what costs us.

**Comment by mortberg** [comment]
@Saizan I can think of two reasons for why this is so bad in Cubical Agda:

1. We are always doing "endpoint correction" in all parametrized HITs right? 

2. We should implement the transp for hcomp in U as in cubicaltt using the optimizations I mentioned a while ago. This should be more efficient than going through transp in Glue.

**Comment by mortberg** [comment]
> > > @3abc you are right. I tried an even more trivial one (that 0-0 = 0-1) and it still hangs.
> > > However, I'm not convinced it's a performance problem than it's a BUG. I don't think the normalization involved is that complex.
> > 
> > 
> > Are you tests the same as:
> > https://github.com/mortberg/cubicaltt/blob/master/examples/hz.ctt#L63
> > ?
> > Computing `0=1` and `1=1` is instant:
> > ```
> > > test01
> > EVAL: false
> > Time: 0m0.089s
> > > test11
> > EVAL: true
> > Time: 0m0.093s
> > ```
> > So this sounds like either a bug or that your proofs are too complicated.
> 
> I believe there is difference between just saying "yes/no" and saying "yes/no" with a reason. The reason is what costs us.

Oh right, can you write a function returning a boolean without expecting the reason? That seems like what we really want anyway.

**Comment by 3abc** [comment]
> I assume in general, if the type of the term is some inductive type, then by canonicity, the normalized form should be very short? Also I guess the process shouldn't take long because the term would be kept short?

Could be. If I define
```Agda
r : {A : Set} → Dec A → ℕ
r (yes p) = 1
r (no ¬p) = 0
```
then it's easy for Agda to compute `r` of those terms. Returns instantly.

**Comment by riaqn** [comment]
@mortberg  https://github.com/agda/cubical/pull/98

**Comment by mortberg** [comment]
> > I assume in general, if the type of the term is some inductive type, then by canonicity, the normalized form should be very short? Also I guess the process shouldn't take long because the term would be kept short?
> 
> Could be. If I define
> 
> ```agda
> r : {A : Set} → Dec A → ℕ
> r (yes p) = 1
> r (no ¬p) = 0
> ```
> 
> then it's easy for Agda to compute `r` of those terms. Returns instantly.

Ok, very good! What might one want to have the normal form of the proof witnesses for? I don't find it too surprising that they are gigantic and hence unusable...

**Comment by riaqn** [comment]
I think we are good to merge now. 

PS. we will soon need some tests to make sure the code works as we expected.

**Comment by mortberg** [comment]
> I think we are good to merge now.
> 
> PS. we will soon need some tests to make sure the code works as we expected.

Sounds good. I restarted the Travis build and once it's green I'll merge this. 

Please make a PR with some tests checking that it all computes as expected

**Comment by riaqn** [comment]
How about we put the tests in the same file with `private`?
```
private
  test0 : Dec→Bool (discreteℤ [ (3 , 5) ] [ (4 , 6) ]) ≡ true
  test0 = refl

  test1 : Dec→Bool (discreteℤ [ (3 , 5) ] [ (4 , 7) ]) ≡ false
  test1 = refl
```

**Comment by mortberg** [comment]
> How about we put the tests in the same file with `private`?
> 
> ```
> private
>   test0 : Dec→Bool (discreteℤ [ (3 , 5) ] [ (4 , 6) ]) ≡ true
>   test0 = refl
> 
>   test1 : Dec→Bool (discreteℤ [ (3 , 5) ] [ (4 , 7) ]) ≡ false
>   test1 = refl
> ```

Exactly what I had in mind. You can even name them `_`

**Comment by riaqn** [comment]
Great. Do you think it should be in this PR, or  should be a new PR?

**Comment by mortberg** [comment]
Just do it in this one

**Pull request merged**